phoenix-core-client/src/main/java/org/apache/phoenix/jdbc/HighAvailabilityPolicy.java

```diff
@@ -44,63 +46,96 @@ public Connection provide(HighAvailabilityGroup haGroup, Properties info,
-                            .Builder(SQLExceptionCode.HA_CLOSED_AFTER_FAILOVER)
-                            .setMessage("Phoenix connection got closed due to failover")
-                            .setHaGroupInfo(haGroup.getGroupInfo().toString()));
-                    LOG.info("Closed all connections to cluster {} for HA group {}",
-                            jdbcZKUrl, haGroup.getGroupInfo());
-                } finally {
-                    if (cqs != null) {
-                        //CQS is closed but it is not invalidated from global cache in PhoenixDriver
-                        //so that any new connection will get error instead of creating a new CQS
-                        LOG.info("Closing CQS after cluster '{}' becomes STANDBY", zkUrl);
-                        cqs.close();
-                        LOG.info("Successfully closed CQS after cluster '{}' becomes STANDBY",
-                                zkUrl);
+
+        @Override
+        void transitRoleRecordRegistry(HighAvailabilityGroup haGroup, ClusterRoleRecord oldRecord,
+                                       ClusterRoleRecord newRecord) throws SQLException {
+            Optional<String> activeUrl = oldRecord.getActiveUrl();
+
+            //Close connections for Active HBase cluster as there is a change in Registry Type
+            if (activeUrl.isPresent()) {
+                LOG.info("Cluster {} has a change in registryType in HA group {}, now closing "
+                        + "all its connections", activeUrl.get(), oldRecord.getRegistryType());
+                closeConnections(haGroup, activeUrl.get(), oldRecord.getRegistryType());
+            } else {
+                LOG.info("None of the cluster in HA Group {} is active", haGroup);
+            }
+        }
+
+        /**
+         * For FAILOVER Policy if there is a change in active url or no new active url then close all connections.
+         * (url1, ACTIVE, url2, STANDBY) --> (url1, ACTIVE, url3, STANDBY) //Nothing is needed
+         * (url1, ACTIVE, url2, STANDBY) --> (url3, ACTIVE, url2, STANDBY) //Active url change close connections
+         * (url1, ACTIVE, url2, STANDBY) --> (url3, ACTIVE, url4, STANDBY) //Active url change close connections
+         * (url1, ACTIVE, url2, STANDBY) --> (url3, ACTIVE, url1, STANDBY)
+         *                          //Here active became standby but other url changed close connections
+         * (url1, OFFLINE, url2, STANDBY) --> (url3, ACTIVE, url2, STANDBY) //Nothing to do as there were no connections
+         * (url1, ACTIVE, url2, STANDBY) --> (url3, OFFLINE, url2, STANDBY) //Closing old connections as no new active url
+         * @param haGroup The high availability (HA) group
+         * @param oldRecord The older cluster role record cached in this client for the given HA group
+         * @param newRecord New cluster role record read from one ZooKeeper cluster znode
+         * @throws SQLException when not able to close connections
```

Comment:
Need to explain what url1 ... url4 stand for. Not sure what the transitions standfor?